### PR TITLE
feat: add ControllerUUID facadeContext/sharedServerContext

### PR DESCRIPTION
**Pull request by Aflynn50**
The controllerUUID does not change accross the lifetime of the controller, therefore, it makes no sense to keep fetching it from the model config service, instead, store it in the apiserver itself and fetch it from there via the facadeContext when needed.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing
~- [ ] [Integration tests](https://github.com/juju/juju/tree/main/tests), with comments saying what you're testing~
~- [ ] [doc.go](https://discourse.charmhub.io/t/readme-in-packages/451) added or updated in changed packages~

### Unit tests
CI Unit tests
### Integration tests
```
juju bootstrap lxd test
juju add-model default
juju deploy ubuntu -n 2
juju destroy-model default
```

## Links

**Jira card:** [JUJU-6260](https://warthogs.atlassian.net/browse/JUJU-6260)

[JUJU-6260]: https://warthogs.atlassian.net/browse/JUJU-6260?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by SimonRichardson — COMMENTED** [review]
We can also get rid of it from the `registerUserHandler`, either by passing in the controllerUUID or taking it from the controller config.

https://github.com/juju/juju/blob/5e3648a9ac91433d9c92077a433a3189627a33d6/apiserver/registration.go#L201C2-L209

**Comment by SimonRichardson** [comment]
Remove it from the following to see what the damage is.

https://github.com/juju/juju/blob/5e3648a9ac91433d9c92077a433a3189627a33d6/apiserver/common/modelmanagerinterface.go#L75

https://github.com/juju/juju/blob/5e3648a9ac91433d9c92077a433a3189627a33d6/apiserver/common/modelmanagerinterface.go#L111

**Comment by Aflynn50** [comment]
I tried removing those two. We can get away with removing `ControllerUUID` from the `ModelManagerBackend` but for the `Model` interface this is passed back into state at some point under an interface that required `ControllerUUID`. I started trying to remove it from there but realised its a bad idea because doing `model.ControllerUUID()` is really the canonical way to do it there.

I've got rid of all the places on where we get the `ControllerUUID` from state but it can be replaced with getting it from the API server instead.

**Comment by SimonRichardson** [comment]
> I started trying to remove it from there but realised its a bad idea because doing model.ControllerUUID() is really the canonical way to do it there.

We need to remove it and It needs to be done at some point. 
A solution is to make it a dependency of the method that is requiring it.

File: apiserver/facade/interface.go [hunk 1]
```diff
@@ -147,6 +147,9 @@ type ModelContext interface {
 	// a good idea; see Resources.
 	ID() string
 
+	// ControllerUUID returns the controllers unique identifier.
```
Comment (manadart) [inline apiserver/facade/interface.go:150]: Missing apostrophe.

File: apiserver/facades/agent/storageprovisioner/register.go [hunk 2]
```diff
@@ -63,5 +63,6 @@ func newFacadeV4(stdCtx context.Context, ctx facade.ModelContext) (*StorageProvi
 		serviceFactory.Storage(registry),
 		ctx.Logger().Child("storageprovisioner"),
 		modelInfo.UUID,
+		ctx.ControllerUUID(),
```
Comment (manadart) [inline apiserver/facades/agent/storageprovisioner/register.go:66]: You can remove the `ControllerConfig` service input, as it's only used to get this.

File: apiserver/facades/client/cloud/register.go [hunk 3]
```diff
@@ -55,7 +55,7 @@ func newFacadeV7(context facade.ModelContext) (*CloudAPI, error) {
 		}
 
 		return service.CredentialValidationContext{
-			ControllerUUID: m.ControllerUUID(),
```
Comment (manadart) [inline apiserver/facades/client/cloud/register.go:58]: Not to do with your change, but we shouldn't even be getting this model (m) out of legacy state.

File: apiserver/registration.go [hunk 4]
```diff
@@ -205,7 +205,7 @@ func (h *registerUserHandler) getSecretKeyLoginResponsePayload(
 	caCert, _ := controllerConfig.CACert()
 	payload := params.SecretKeyLoginResponsePayload{
 		CACert:         caCert,
-		ControllerUUID: st.ControllerUUID(),
+		ControllerUUID: controllerConfig.ControllerUUID(),
```
Comment (manadart) [inline apiserver/registration.go:208]: There is a chain through `httpContext` -> `Server` -> `sharedServerContext`. I think you should be able to get it from there without even needing the service.

**Review by manadart — COMMENTED** [review]
Looking good.

File: apiserver/registration.go (hunk 4 above)
Comment (Aflynn50) [reply apiserver/registration.go:208]: I saw that, but went with the config since we use it anyway here for the `CACert` and the chain is pretty ugly.

Comment (manadart) [reply apiserver/registration.go:208]: I missed that we already had it. This is fine.

**Review by manadart — APPROVED** [review]
QA steps affecting the API server should have at least some nominal bootstrap + deploy (which I've done).

Otherwise the reviewer cannot assume it's even been run up.

File: apiserver/common/modelmanagerinterface.go [hunk 5]
```diff
@@ -108,6 +107,9 @@ type Model interface {
 	MigrationMode() state.MigrationMode
 	Name() string
 	UUID() string
+	// TODO(aflynn): ControllerUUID is only here because the EnvironConfigGetter
```
[the remark this answers is not in the document]
Comment (Aflynn50) [reply apiserver/common/modelmanagerinterface.go:110]: Nice, that'll be good to pull out.

**Comment by Aflynn50** [comment]
> QA steps affecting the API server should have at least some nominal bootstrap + deploy (which I've done).
> 
> Otherwise the reviewer cannot assume it's even been run up.

Ah, I had actually done some, I just forgot to put it in the PR. I've updated it with what I did.

**Pull request merged**